Update full quiz story docs

mocca102 opened the pull request:
- Updated callbacks key to include all callbacks
- New docs layout
- Update callbacks, resultPageOptions and resultCardOptions

mocca102 commented on .storybook/main.js:
```diff
@@ -27,4 +27,19 @@ module.exports = {
     ...config,
     LOGGER: true,
   }),
+  typescript: {
```
Options to autogenerate types from typescript
It works for direct props of the component and not nested types. Thus the usage of argTypes as well to document the nested types

mocca102 commented on .storybook/preview.jsx:
```diff
@@ -22,13 +22,7 @@ export const parameters = {
-        <Stories includePrimary />
-      </>
-    )
+    toc: true,
```
New navigation for page table of contents 
<img width="1726" alt="Screenshot 2023-09-10 at 10 10 22 PM" src="https://github.com/Constructor-io/constructorio-ui-quizzes/assets/58053149/adec0ffb-95f1-41da-90af-1900aee23587">

mocca102 replied:
This is part of the updated storybook version. I removed the update changes since it created a big PR and created a separate story for storybook update

mocca102 commented on package.json:
```diff
@@ -66,14 +66,14 @@
-    "@storybook/addon-interactions": "^7.0.11",
-    "@storybook/addon-links": "7.0.11",
-    "@storybook/addon-docs": "7.0.11",
+    "@storybook/addon-a11y": "^7.4.0",
```
Updated storybook

mocca102 commented on src/types.ts:
```diff
@@ -59,18 +56,22 @@ export type QuestionWithAnswer = Question & {
 export interface Callbacks {
-  onQuizNextQuestion?: OnQuizNextQuestion;
+  onQuizNextQuestion?: QuizResultsEventsProps.OnQuizNextQuestion;
+  onQuizResultsLoaded?: QuizResultsEventsProps.OnQuizResultsLoaded;
```
Move the rest of callbacks from resultsPageOptions to callbacks

mocca102 commented on src/types.ts:
```diff
@@ -59,18 +56,22 @@ export type QuestionWithAnswer = Question & {
   quizVersionId?: string;
-  resultsPageOptions: ResultsPageOptions;
+  resultsPageOptions?: ResultsPageOptions;
+  resultCardOptions?: ResultCardOptions;
```
Added ResultCardOptions to include result card related props

mocca102 commented on src/hooks/useConsoleErrors.ts:
```diff
@@ -2,23 +2,18 @@ import { useEffect } from 'react';
     }
 
-    if (!resultsPageOptions || Object.keys(resultsPageOptions).length === 0) {
+    if (!callbacks?.onAddToCartClick) {
```
Removed resultsPageOptions since they're only required for the component interface and not for hooks.
Added to the docs though

mocca102 commented on src/hooks/useConsoleErrors.ts:
```diff
@@ -2,23 +2,18 @@ import { useEffect } from 'react';
       console.error('quizId is a required field of type string');
     }
 
-    if (!resultsPageOptions || Object.keys(resultsPageOptions).length === 0) {
```
Removed resultsPageOptions since they're only required for the component interface and not for hooks.
Added to the docs though

mocca102 commented on src/types.ts:
```diff
@@ -39,14 +39,11 @@ export namespace QuizResultsEventsProps {
 }
 
-export interface ResultsPageOptions extends ResultCardProps {
+export interface ResultsPageOptions {
```
Removed callbacks

mocca102 commented on .github/workflows/run-license-check.yml:
```diff
@@ -20,6 +20,6 @@ jobs:
           node-version: ${{ matrix.node }}
       - name: Install dependencies
-        run: npm install
+        run: npm ci
```
For consistency

mocca102 commented on .storybook/preview.jsx:
```diff
@@ -21,15 +21,6 @@ export const parameters = {
   previewTabs: {
     'storybook/docs/panel': { index: -1 }
   },
-  docs: {
```
Using auto generated docs with some tweaks instead

esezen commented on src/hooks/useConsoleErrors.ts:
```diff
@@ -2,23 +2,18 @@ import { useEffect } from 'react';
+    if (!callbacks?.onAddToCartClick) {
       // eslint-disable-next-line no-console
-      console.error('resultsPageOptions is a required field of type object');
+      console.error('callbacks?.onAddToCartClick is a required field of type function');
```
`?` is probably here because of copy paste. Can we change it to `callbacks.onAddToCartClick`?

esezen commented on src/stories/Quiz/argTypes.ts:
```diff
@@ -1,32 +1,58 @@
+  resultCardOptions: {
+    description: `Result card configuration options
+      (Required for the component interface and has to match the keys in the API response to be correctly rendered)
+      \`resultCardRegularPriceKey: string\`
```
Can we add a `\n` to this line to force it to go to the next line?

esezen reviewed (commented):
This is looking great. I like the new format a lot. Let's get another review from the team since this is a big PR and coordinate the release with the affected parties.

stanlp1 reviewed (changes requested):
The results page for "Render in a small container" "Provide CIO client instance", "Pass callbacks" are broken for me and don't render 🤔

mocca102 commented:
> The results page for "Render in a small container" "Provide CIO client instance", "Pass callbacks" are broken for me and don't render 🤔

@stanlp1 resultCardOptions was missing. Passed it as a parameter 

Also we have this in the docs to indicate that it's required 
"Result card configuration options (Required for the component interface and has to match the keys in the API response to be correctly rendered)"

jjl014 reviewed (commented):
Nice, this is looking great to me!

I leave a few comments and have some questions.

The way that the description is currently set up is a bit hard to read as there's various text that are wrapped in backticks and the type that gets appended at the end. I created a [PR here](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/134) on a potential solution, but I'm not sure if it's a valid approach. 🤔 

Let me know what you think!

mocca102 commented on src/components/ResultContainer/ResultContainer.tsx:
```diff
@@ -3,58 +3,63 @@ import RedoButton from '../RedoButton/RedoButton';
 import Results from '../Results/Results';
 import Spinner from '../Spinner/Spinner';
 
 export interface IResultContainerProps {
```
No particular reason the letter `I` stands for interface just to indicate that it's an interface. I just was not consistent

mocca102 commented on src/hooks/useQuizEvents/index.ts:
```diff
@@ -26,10 +26,7 @@ const useQuizEvents: UseQuizEvents = (quizOptions, cioClient, quizState) => {
-
-  const { onAddToCartClick, onQuizResultClick, onQuizResultsLoaded, onAddToFavoritesClick } =
-    resultsPageOptions;
+  const { callbacks } = quizOptions;
```
Callback is optional so I couldn't destructure it without an If condition which meant doing let first for all of the callbacks so I figured accessing them like `callbacks?.[function]` was easier

jjl014 replied:
Ooo, got it. I think the following would work, but feel free to ignore this. 😄 

```
const { callbacks } = quizOptions || {};
const { onAddToCartClick, ... } = callbacks || {};
```

mocca102 replied:
nice trick

mocca102 commented:
> Nice, this is looking great to me!
> 
> I left a few comments and a question.
> 
> Is there anyway to break the nested options out of the description? The way that the description is currently set up is a bit hard to read as there's various text that are wrapped in backticks and the type that gets appended at the end. I created a [PR here](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/134) on a potential solution, but I'm not sure if it's a valid approach. 🤔
> 
> Let me know what you think!
> 
> cc @esezen

@jjl014 I'm afraid this won't work. The approach in the draft you shared for this snippet for example 
```
 onAddToCartClick: {
    description: '...',
    table: {
      subcategory: 'callbacks',
    },
  }

```

Storybook treats `onAddToCartClick` as a direct prop to CioQuiz and this creates problems with type matching when we pass the argTypes to the story config since `onAddToCartClick` isn't a direct prop of CioQuiz but rather a prop on `callbacks`

@jjl014

mocca102 commented:
@jjl014 Merged your draft PR and added some tweak. Added callbacks.[function] instead of just function so it's clear to users that it's a property on callbacks. Let me know what you think about that.

Also hid the controls that aren't useful and left the ones we usually use.

jjl014 commented on src/stories/Quiz/Component/ComponentStories.stories.tsx:
```diff
@@ -1,78 +1,50 @@
+import type { Meta } from '@storybook/react';
 import CioQuiz from '../../../components/CioQuiz';
-import { argTypes } from '../argTypes';
+import { argTypes, docsControls, storiesControls } from '../argTypes';
```
Seems like we can remove `docsControl` from here as it's not used

jjl014 commented on src/stories/Quiz/Hooks/index.stories.tsx:
```diff
@@ -3,7 +3,7 @@ import MyQuiz, { addHookStoryCode } from './index';
+import { argTypes, docsControls, storiesControls } from '../argTypes';
 
 export default {
   title: 'Quiz/Hooks',
```
Not sure if this is the right place, but it seems like the control for `primaryColor` isn't working here. Tried without quotes, with quotes, with single quotes. 🤔 

Also, why do some of the descriptions have a `-` and some don't? I think they were showing that the type was in the Component Basic Usage story, but it seems like those aren't shown here.

![Screenshot 2023-10-05 at 5 25 20 PM](https://github.com/Constructor-io/constructorio-ui-quizzes/assets/7194266/2caf1544-5208-409c-9a30-6ded7941fc9f)

Also, this is just me being anal retentive, but could we be a bit more consistent with the capitalizations here?
* `import` => `Import`
* `use` => `Use`
* And I'll leave capitalizing the sub-bullets (i.e. `state management`) to your best judgement

![Screenshot 2023-10-05 at 5 30 23 PM](https://github.com/Constructor-io/constructorio-ui-quizzes/assets/7194266/9370cef4-0edb-4c4e-82d1-881e09adaf22)

mocca102 replied:
> Not sure if this is the right place, but it seems like the control for primaryColor isn't working here
The primary colors were not used in the hooks example. I added it in the example but it's not very user friendly.

> Also, why do some of the descriptions have a - and some don't? I think they were showing that the type was in the Component Basic Usage story, but it seems like those aren't shown here.
We have another PR for the hooks docs. But removed the `-` for now

jjl014 commented on src/stories/Quiz/Examples/index.stories.tsx:
```diff
@@ -1,13 +1,16 @@
 
 export default {
   title: 'Quiz/Specific Examples',
   component: CioQuiz,
```
It looks like the controls on the `Specific Examples/Docs` page doesn't seem to be doing anything. Is it possible to fix them or should we disable the controls?

mocca102 replied:
Better to be removed

jjl014 commented on src/stories/Quiz/Component/ComponentStories.mdx:
```diff
@@ -0,0 +1,126 @@
+<Canvas>
+  <Story of={ComponentStories.BasicUsage} />
+</Canvas>
+<h2>Quiz Controls</h2>
```
Would it make sense to change this to be `Example Quiz Controls` or `Sample Quiz Controls` since it only shows a subset of the possible config options?

mocca102 replied:
In storybook Controls means options you configured to be controlled not necessarily all config options. I named it to just controls

jjl014 commented on src/stories/Quiz/Component/ComponentStories.mdx:
```diff
@@ -0,0 +1,126 @@
+- [Provide CIO Client Instance](#provide-cio-client-instance)
+- [Handle Favorites On Results Page](#handle-favorites-on-results-page)
+
+<h2>Quiz Props</h2>
```
I think a couple of the fields at the top do have defaults right?
* `enableHydration` - defaults to `false`?
* `primaryColor` - seems like this might have a default of a blue color? I could be wrong

jjl014 commented on src/stories/Quiz/tests/index.stories.tsx:
```diff
@@ -20,33 +21,25 @@ export default {
 // eslint-disable-next-line storybook/prefer-pascal-case
 export const e2eInteractionTest = ComponentTemplate.bind({});
-e2eInteractionTest.args = { apiKey, quizId, resultsPageOptions, sessionStateOptions };
+e2eInteractionTest.args = {
```
Hopefully, I'm leaving this comment in the right place.

Would it be possible to move the `Full Quiz e2e tests` story to the end? Also, moving Questions and Results right before it, but also towards the end.

Please reference
![Screenshot 2023-10-05 at 5 32 45 PM](https://github.com/Constructor-io/constructorio-ui-quizzes/assets/7194266/b211e7b1-1dfd-4cf9-b9f9-095c7db3cd10)
 my poor attached drawing.

jjl014 reviewed (commented):
Nice, I'm digging these changes. Thanks for making the updates!

I left a couple of comments. Also, is there anything we added previously that can be removed now that we're going with this approach?

Would also love to get some additional eyes from the @Constructor-io/customer-integrations team and get their thoughts as well.

Thanks!

mocca102 commented on src/stories/Quiz/Examples/index.stories.tsx:
```diff
@@ -1,16 +1,8 @@
-  component: CioQuiz,
-  argTypes,
-  parameters: {
-    controls: storiesControls,
```
Removed unnecessary controls

mocca102 commented on src/stories/Quiz/Hooks/index.stories.tsx:
```diff
@@ -15,21 +21,23 @@ export default {
       description: {
         component: hookDescription,
       },
-      controls: docsControls,
```
This is temporary until we push the hooks docs

mocca102 commented on src/stories/Quiz/Hooks/index.tsx:
```diff
@@ -106,6 +107,7 @@ export default function HooksTemplate(args) {
       if (currentQuestion.isOpenQuestion) {
         return (
           <div className='cio-quiz'>
+            <style>.cio-quiz {convertPrimaryColorsToString(primaryColorStyles)}</style>
```
`primaryColorStyles` wasn't being used in the hooks example. But this isn't a friendly customer interface so we still need to handle this in a separate PR maybe

jjl014 reviewed (approved):
Love all the changes. Thank you for putting up with my asks. ❤️ This look fantastic to me and I can't wait to get these changes out! 🚀 

I would love to get one other review from the team, but otherwise this looks amazing! Thanks again!

esezen reviewed (approved):
Love these changes! Thanks for working on this @mocca102 and thanks for the reviews @stanlp1 @jjl014 🚀 

I think we are good to merge this in, let's coordinate the `callback` changes with the affected parties to make sure we don't cause any issues
